Deferred/KSPromise.h

```diff
@@ -2,27 +2,25 @@
 #import "KSCancellable.h"
 #import "KSNullabilityCompat.h"
 
-@class KSPromise;
-
 NS_ASSUME_NONNULL_BEGIN
 
-typedef __nonnull id(^promiseValueCallback)(id value);
-typedef __nonnull id(^promiseErrorCallback)(NSError *error);
-typedef void(^deferredCallback)(KSPromise *p);
-
 FOUNDATION_EXPORT NSString *const KSPromiseWhenErrorDomain;
 FOUNDATION_EXPORT NSString *const KSPromiseWhenErrorErrorsKey;
 FOUNDATION_EXPORT NSString *const KSPromiseWhenErrorValuesKey;
 
-@interface KSPromise : NSObject<KSCancellable>
-@property (strong, nonatomic, readonly, nullable) id value;
+@interface KSPromise<ObjectType> : NSObject<KSCancellable>
+typedef __nonnull id(^promiseValueCallback)(ObjectType value);
+typedef __nonnull id(^promiseErrorCallback)(NSError *error);
+typedef void(^deferredCallback)(KSPromise<ObjectType> *p);
+
+@property (strong, nonatomic, readonly, nullable) ObjectType value;
 @property (strong, nonatomic, readonly, nullable) NSError *error;
 @property (assign, nonatomic, readonly) BOOL fulfilled;
 @property (assign, nonatomic, readonly) BOOL rejected;
 @property (assign, nonatomic, readonly) BOOL cancelled;
 
 + (KSPromise *)when:(NSArray *)promises;
-- (KSPromise *)then:(nullable promiseValueCallback)fulfilledCallback error:(nullable promiseErrorCallback)errorCallback;
+- (KSPromise *)then:(nullable id(^)(ObjectType value))fulfilledCallback error:(nullable promiseErrorCallback)errorCallback;
```

Comment:
Should `then:error:` actually return a `KSPromise<ObjectType> *` ? Or does that not work?